external_ip_mapper.go

```diff
@@ -0,0 +1,129 @@
+package ice
+
+import (
+	"net"
+	"strings"
+)
+
+func validateIPString(ipStr string) (net.IP, bool, error) {
+	ip := net.ParseIP(ipStr)
+	if ip == nil {
+		return nil, false, ErrInvalidNAT1To1IPMapping
+	}
+	return ip, (ip.To4() != nil), nil
+}
+
+type externalIPMapper struct {
+	ipv4Map            map[string]net.IP
+	ipv6Map            map[string]net.IP
+	isLocIPv4Specified bool // if true, ipv4Map has 0 or 1 external IPv4
+	isLocIPv6Specified bool // if true, ipv6Map has 0 or 1 external IPv6
+	candidateType      CandidateType
+}
+
+func newExternalIPMapper(candidateType CandidateType, ips []string) (*externalIPMapper, error) {
+	if len(ips) == 0 {
+		return nil, nil
+	}
+	if candidateType == CandidateTypeUnspecified {
+		candidateType = CandidateTypeHost // defaults to host
+	} else if candidateType != CandidateTypeHost && candidateType != CandidateTypeServerReflexive {
+		return nil, ErrUnsupportedNAT1To1IPCandidateType
+	}
+
+	m := &externalIPMapper{
+		ipv4Map:       map[string]net.IP{},
+		ipv6Map:       map[string]net.IP{},
+		candidateType: candidateType,
+	}
+
+	for _, extIPStr := range ips {
+		ipPair := strings.Split(extIPStr, "/")
+		if len(ipPair) == 0 || len(ipPair) > 2 {
+			return nil, ErrInvalidNAT1To1IPMapping
+		}
+
+		extIP, isExtIPv4, err := validateIPString(ipPair[0])
+		if err != nil {
+			return nil, err
+		}
+		if len(ipPair) == 1 {
```

Comment:
sorry if this is not valid, I should test this myself but I want to give you a review before going to bed:

Let's say the first mapping is a single ip `1.2.3.4` and the second one is a pair `1.2.3.5/10.1.5.26`. What your ipv4Map will contain, if I'm reading the code correctly, is: 

```
{ 
   "*": "1.2.3.4",
   "10.1.5.26": "1.2.3.5"
}
```

and `isLocIPv4Specified` will be true. 

Now, something is not right if that's the case, because the `findExternalIP` will ignore `*` if `isLocIPv4Specified` is true

Reply:
You are such an awesome reviewer! I did not have the test for that case, so I added one, which ended up with the exact result you predicted! And that is not what I expected. I will fix that.

Reply:
Glad it helped!, I was thinking about how can we simplify the code in order to catch this kind of errors and I came up with a little refactor: https://github.com/pion/ice/commit/3992531489c3aaead524f4145bfab86ce44c7f7d let me know what you think (feel free to pick just some pieces or nothing at all). Anyway I will aprove the PR right now

Reply:
Oh, I love it. Thank you!  I merged your commit, then did a bit of naming and trivial code optimization.